src/librustdoc/formats/cache.rs

```diff
@@ -131,44 +130,21 @@ struct CacheBuilder<'a, 'tcx> {
 }
 
 impl Cache {
-    crate fn from_krate<'tcx>(
-        render_info: RenderInfo,
-        document_private: bool,
+    crate fn new(access_levels: AccessLevels<DefId>, document_private: bool) -> Self {
+        Cache { access_levels, document_private, ..Cache::default() }
+    }
+
+    crate fn populate(
+        &mut self,
+        mut krate: clean::Crate,
+        tcx: TyCtxt<'_>,
         extern_html_root_urls: &BTreeMap<String, String>,
         dst: &Path,
-        mut krate: clean::Crate,
-        tcx: TyCtxt<'tcx>,
-    ) -> (clean::Crate, Cache) {
+    ) -> clean::Crate {
         // Crawl the crate to build various caches used for the output
-        let RenderInfo {
-            inlined: _,
-            external_paths,
-            exact_paths,
-            access_levels,
-            deref_trait_did,
-            deref_mut_trait_did,
-            owned_box_did,
-            ..
-        } = render_info;
-
-        let external_paths =
-            external_paths.into_iter().map(|(k, (v, t))| (k, (v, ItemType::from(t)))).collect();
-
-        let mut cache = Cache {
-            external_paths,
-            exact_paths,
-            parent_is_trait_impl: false,
-            stripped_mod: false,
-            access_levels,
-            crate_version: krate.version.take(),
-            document_private,
-            traits: krate.external_traits.replace(Default::default()),
-            deref_trait_did,
-            deref_mut_trait_did,
-            owned_box_did,
-            masked_crates: mem::take(&mut krate.masked_crates),
-            ..Cache::default()
-        };
```

Comment:
Just to make sure I understand, these initializations can be removed because:

- some of them (`parent_is_trait_impl`, `stripped_mod`) are handled by the use of `default()` in `Cache::new()`
- some of them (`document_private`, `access_levels`) are handled by explicit initialization in `Cache::new()`
- some of them (`traits`, `masked_crates`) are initialized using assignment later on in the method
- the rest are initialized directly by users of `Cache`

right?

Reply:
Yes, that sounds right.